core/slim_globals.cpp

```diff
@@ -1647,17 +1647,37 @@ void SLiM_ConfigureContext(void)
 
 const std::string gSLiM_tsk_edge_metadata_schema = "";
 const std::string gSLiM_tsk_site_metadata_schema = "";
 
 const std::string gSLiM_tsk_mutation_metadata_schema =
 R"V0G0N({"$schema":"http://json-schema.org/schema#","additionalProperties":false,"codec":"struct","description":"SLiM schema for mutation metadata.","examples":[{"mutation_list":[{"mutation_type":1,"nucleotide":3,"selection_coeff":-0.2,"slim_time":243,"subpopulation":0}]}],"properties":{"mutation_list":{"items":{"additionalProperties":false,"properties":{"mutation_type":{"binaryFormat":"i","description":"The index of this mutation's mutationType.","index":1,"type":"integer"},"nucleotide":{"binaryFormat":"b","description":"The nucleotide for this mutation (0=A , 1=C , 2=G, 3=T, or -1 for none)","index":5,"type":"integer"},"selection_coeff":{"binaryFormat":"f","description":"This mutation's selection coefficient.","index":2,"type":"number"},"slim_time":{"binaryFormat":"i","description":"The SLiM tick counter when this mutation occurred.","index":4,"type":"integer"},"subpopulation":{"binaryFormat":"i","description":"The ID of the subpopulation this mutation occurred in.","index":3,"type":"integer"}},"required":["mutation_type","selection_coeff","subpopulation","slim_time","nucleotide"],"type":"object"},"noLengthEncodingExhaustBuffer":true,"type":"array"}},"required":["mutation_list"],"type":"object"})V0G0N";
 
+// BCH 12/10/2024: Removed the type field, and changed the treatment of is_null.  We have a
+// tricky problem here, which is that is_null is now variable-length and there is no count.
+// The number of byte (uint8_t) entries in is_null depends on the number of chromosomes in
+// the full set of tree sequences, because the node metadata has to contain flags (bits) for
+// every chromosome, not just for the chromosome represented by this file.  So we deduce the
+// length of is_null from that, but it is variable and has no count associated with it in
+// the metadata here.  I think this is actually not allowed in JSON Schema, understandably.
+// The schema here provides 8 bits, so it suffices for up to an 8-chromosome model.  We can
+// have up to 256 chromosomes, so we would have up to 32 uint8_t bytes of is_null data here,
+// in that extreme case.  It is very ugly, but I think the solution might be to write out
+// a different version of the metadata schema depending on the number of bytes used.  In
+// other words, if 7 bytes of is_null data are needed (for 49-56 chromosomes), we'd write
+// out a version of the schema that specifies 7 bytes of is_null data, named is_null1 to
+// is_null7, I suppose.  This effectively puts the count into the schema itself.  Peter,
+// do you see a better option than this?  I'd really prefer not to add a count, since it's
+// completely redundant.
```

Comment:
Hm - well, looking at [the tskit metadata docs for binary encoding](https://tskit.dev/tskit/docs/latest/metadata.html#struct), I agree - arrays have to carry their length along with them. Hm - but, how about using a [string](https://tskit.dev/tskit/docs/latest/metadata.html#binaryformat)? The schema would still depend on the number of chromosomes, since you'd have like `"binaryFormat" : "10p"`, where `10` is the number of char's we need for that many chromosomes, but seems a lot nicer than some sequence of variables?

We could extend `tskit`'s metadata schema parsing to make this easier if necessary; but, we're still limited by the [`struct` module](https://docs.python.org/3/library/struct.html), and I'm not seeing a better option there than like `10p`.

Reply:
Hm - actually the `p` specification encodes the length, so that doesn't help, but something like that (`10c`?) might?
![Screenshot from 2024-12-20 12-01-46](https://github.com/user-attachments/assets/73364323-19a2-467c-9d40-91b1f824198d)

Reply:
Yes - I think my proposal is to do like `"binaryFormat" : "10c"` where `10` is how many bytes you want.

Reply:
OK – and write a different schema (`1c` vs. `10c` etc.) based on how many bytes are needed.  I'll have to review B vs. c, but isn't c a signed char, and B is a uint8?  Unsigned seems better; having the highest bit interpreted as a sign bit just makes things that much more bug-prone.

Reply:
That's the idea, you know better than I about the precise type.

Reply:
OK, I've done this.  I'm going to "resolve" these conversations as I go, so I can easily see what is left for me to do.  Resolving this now.